packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -364,13 +364,15 @@ export default createRule<Options, MessageId>({
       '!==',
     ]);
     function checkIfBinaryExpressionIsNecessaryConditional(
-      node: TSESTree.BinaryExpression,
+      node: TSESTree.Node,
+      {
+        left,
+        right,
+        operator,
+      }: Pick<TSESTree.BinaryExpression, 'left' | 'right' | 'operator'>,
     ): void {
-      if (!BOOL_OPERATORS.has(node.operator)) {
```

Comment:
(nit) - i'm not sure i agree with moving this check outside the function. The function's title implies that it can accept any binary expression, so it would seem correct for it to validate first that the expression is a conditional, unless the `TSESTree.BinaryExpression` type itself could guarantee that the binary expression is in `BOOL_OPERATORS`.

Reply:
Rather than adding a runtime check that would run unnecessarily for `SwitchCase`s, I instead
- kept the runtime check to continue only checking on `BinaryExpression`s
- used TypeScript to ensure only valid operators are passed into the function (since why use runtime checks when we can use type checks? 🤷‍♂️ )
- renamed the function to be consistent with the existing `BOOL_OPERATORS` constant and make clear that it is only for a subset of `BinaryExpression`s

Reply:
Yep, using the type system to enforce it is good, too. Nice 👍